Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialTypeSourceData.cpp

```diff
@@ -112,89 +127,351 @@ namespace AZ
+            return toPropertyGroupList.back().get();
+        }
+
+        MaterialTypeSourceData::PropertyDefinition* MaterialTypeSourceData::PropertyGroup::AddProperty(AZStd::string_view name)
+        {
+            auto propertyIter = AZStd::find_if(m_properties.begin(), m_properties.end(), [name](const AZStd::unique_ptr<PropertyDefinition>& existingProperty)
+                {
+                    return existingProperty->GetName() == name;    
+                });
+
+            if (propertyIter != m_properties.end())
+            {
+                AZ_Error("Material source data", false, "PropertyGroup '%s' already contains a property named '%.*s'", m_name.c_str(), AZ_STRING_ARG(name));
+                return nullptr;
+            }
+            
+            auto propertyGroupIter = AZStd::find_if(m_propertyGroups.begin(), m_propertyGroups.end(), [name](const AZStd::unique_ptr<PropertyGroup>& existingPropertyGroup)
+                {
+                    return existingPropertyGroup->m_name == name;
+                });
+
+            if (propertyGroupIter != m_propertyGroups.end())
+            {
+                AZ_Error("Material source data", false, "Property name '%.*s' collides with a PropertyGroup of the same name", AZ_STRING_ARG(name));
+                return nullptr;
+            }
+
+            if (!MaterialPropertyId::IsValidName(name))
+            {
+                AZ_Error("Material source data", false, "'%.*s' is not a valid identifier", AZ_STRING_ARG(name));
+                return nullptr;
+            }
+
+            m_properties.emplace_back(AZStd::make_unique<PropertyDefinition>(name));
+            return m_properties.back().get();
+        }
+
+        MaterialTypeSourceData::PropertyGroup* MaterialTypeSourceData::PropertyGroup::AddPropertyGroup(AZStd::string_view name)
+        {
+            auto iter = AZStd::find_if(m_properties.begin(), m_properties.end(), [name](const AZStd::unique_ptr<PropertyDefinition>& existingProperty)
+                {
+                    return existingProperty->GetName() == name;
+                });
+
+            if (iter != m_properties.end())
+            {
+                AZ_Error("Material source data", false, "PropertyGroup name '%.*s' collides with a Property of the same name", AZ_STRING_ARG(name));
+                return nullptr;
+            }
+
+            return AddPropertyGroup(name, m_propertyGroups);
+        }
+        
+        MaterialTypeSourceData::PropertyGroup* MaterialTypeSourceData::AddPropertyGroup(AZStd::string_view propertyGroupId)
+        {
+            AZStd::vector<AZStd::string_view> splitPropertyGroupId = SplitId(propertyGroupId);
+
+            if (splitPropertyGroupId.size() == 1)
+            {
+                return PropertyGroup::AddPropertyGroup(propertyGroupId, m_propertyLayout.m_propertyGroups);
+            }
+
+            PropertyGroup* parentPropertyGroup = FindPropertyGroup(splitPropertyGroupId[0]);
```

Comment:
should this check if `splitPropertyGroupId` is not empty before accessing its first element?